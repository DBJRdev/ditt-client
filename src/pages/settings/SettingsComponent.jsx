import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import React from 'react';
import {
  Button,
  Modal,
  TextArea,
  TextField,
} from 'react-ui';
import { withNamespaces } from 'react-i18next';
import Layout from '../../components/Layout';
import {
  createDate,
  toDayMonthYearFormat,
  toMomentDateTimeFromDayMonth,
} from '../../services/dateTimeService';
import { validateSupportedYear } from '../../services/validatorService';
import styles from './Settings.scss';

class SettingsComponent extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      formData: {
        holidays: '',
        year: null,
      },
      formValidity: {
        elements: {
          holidays: null,
          year: null,
        },
        isValid: false,
      },
      openedYears: [],
      showAddYearDialog: false,
      showEditYearDialog: false,
    };

    this.changeHandler = this.changeHandler.bind(this);
    this.openAddYearDialog = this.openAddYearDialog.bind(this);
    this.closeAddYearDialog = this.closeAddYearDialog.bind(this);
    this.closeEditYearDialog = this.closeEditYearDialog.bind(this);
    this.addHandler = this.addHandler.bind(this);
    this.editHandler = this.editHandler.bind(this);
    this.renderYearsAndHolidaysBox = this.renderYearsAndHolidaysBox.bind(this);
  }

  componentDidMount() {
    this.props.fetchConfig();
  }

  changeHandler(e) {
    const eventTarget = e.target;

    this.setState((prevState) => {
      const formData = Object.assign({}, prevState.formData);
      formData[eventTarget.id] = eventTarget.value;

      return { formData };
    });
  }

  changeOpenedYearHandler(supportedYear) {
    const { openedYears } = this.state;
    const changedOpenedYears = [...openedYears];

    const foundIndex = openedYears.indexOf(supportedYear);

    if (foundIndex === -1) {
      changedOpenedYears.push(supportedYear);
    } else {
      changedOpenedYears.splice(foundIndex, 1);
    }

    this.setState({ openedYears: changedOpenedYears });
  }

  addHandler() {
    const {
      config,
      saveConfig,
      t,
    } = this.props;
    const { formData } = this.state;
    const supportedYears = config.get('supportedYears').toJS();
    const supportedHolidays = config.get('supportedHolidays').toJS();

    const formValidity = validateSupportedYear(t, formData, supportedYears, true);

    this.setState({ formValidity });

    if (formValidity.isValid) {
      supportedYears.push(formData.year);

      if (formData.holidays !== null) {
        formData.holidays.split('\n').forEach((line) => {
          if (line.trim() === '') {
            return;
          }

          const holiday = toMomentDateTimeFromDayMonth(line);
          supportedHolidays.push(
            createDate(formData.year, holiday.get('month'), holiday.get('date'))
          );
        });
      }

      saveConfig({
        supportedHolidays,
        supportedYears,
      }).then(this.closeAddYearDialog);
    }
  }

  editHandler() {
    const {
      config,
      saveConfig,
      t,
    } = this.props;
    const { formData } = this.state;
    const supportedYears = config.get('supportedYears').toJS();
    const supportedHolidays = config.get('supportedHolidays').toJS()
      .filter(holiday => holiday.year() !== formData.year);

    const formValidity = validateSupportedYear(t, formData, supportedYears, false);

    this.setState({ formValidity });

    if (formValidity.isValid) {
      if (formData.holidays !== null) {
        formData.holidays.split('\n').forEach((line) => {
          if (line.trim() === '') {
            return;
          }

          const holiday = toMomentDateTimeFromDayMonth(line);
          supportedHolidays.push(
            createDate(formData.year, holiday.get('month'), holiday.get('date'))
          );
        });
      }

      saveConfig({
        supportedHolidays,
        supportedYears,
      }).then(this.closeEditYearDialog);
    }
  }

  openAddYearDialog() {
    this.setState({
      showAddYearDialog: true,
    });
  }

  closeAddYearDialog() {
    this.setState({ showAddYearDialog: false });
    this.resetForm();
  }

  openEditYearDialog(year) {
    const { config } = this.props;
    const holidays = config.get('supportedHolidays')
      .filter(holiday => holiday.year() === year)
      .map(holiday => holiday.format('MM.DD.'))
      .reduce((acc, holiday) => `${acc}\n${holiday}`);

    this.setState({
      formData: {
        holidays,
        year,
      },
      showEditYearDialog: true,
    });
  }

  closeEditYearDialog() {
    this.setState({ showEditYearDialog: false });
    this.resetForm();
  }

  resetForm() {
    this.setState({
      formData: {
        holidays: '',
        year: null,
      },
      formValidity: {
        elements: {
          holidays: null,
          year: null,
        },
        isValid: false,
      },
    });
  }

  renderYearsAndHolidaysBox(supportedYear) {
    const {
      config,
      t,
    } = this.props;
    const { openedYears } = this.state;
    const isOpened = openedYears.includes(supportedYear);

    const holidays = config.get('supportedHolidays').toJS()
      .filter(holiday => holiday.year() === supportedYear)
      .map(holiday => toDayMonthYearFormat(holiday));

    return (
      <div
        className={styles.yearsAndHolidaysHeader}
        key={supportedYear}
      >
        <div className={styles.yearsAndHolidaysHeaderBox}>
          <h3 className={styles.yearTitle}>
            {supportedYear}
          </h3>
          <div className={styles.yearButtonsWrapper}>
            <div className={styles.yearEditButtonWrapper}>
              <Button
                clickHandler={() => this.openEditYearDialog(supportedYear)}
                icon="edit"
                label={t('settings:action.editYear')}
                isLabelVisible={false}
                priority="default"
              />
            </div>
            <Button
              clickHandler={() => this.changeOpenedYearHandler(supportedYear)}
              icon={isOpened ? 'expand_less' : 'expand_more'}
              label=""
              isLabelVisible={false}
              priority="default"
            />
          </div>
        </div>
        {isOpened && (
          <div className={styles.yearsAndHolidaysContentBox}>
            <p className={styles.holidaysTitle}>
              {t('config:element.holidays')}
              :
            </p>
            <p>
              {holidays.map(holiday => (
                <span key={holiday}>
                  {holiday}
                  <br />
                </span>
              ))}
              {holidays.length === 0 && (
                <span>-</span>
              )}
            </p>
          </div>
        )}
      </div>
    );
  }

  renderAddYearModal() {
    const {
      isPosting,
      t,
    } = this.props;

    return (
      <Modal
        actions={[
          {
            clickHandler: this.addHandler,
            label: t('general:action.save'),
            loading: isPosting,
          },
        ]}
        closeHandler={this.closeAddYearDialog}
        title={t('settings:title.addYear')}
        translations={{ close: t('general:action.close') }}
      >
        <form>
          <TextField
            changeHandler={this.changeHandler}
            error={this.state.formValidity.elements.year}
            fieldId="year"
            label={t('config:element.year')}
            required
            type="number"
            value={this.state.formData.year || ''}
          />
          <TextArea
            changeHandler={this.changeHandler}
            error={this.state.formValidity.elements.holidays}
            fieldId="holidays"
            label={t('config:element.holidays')}
            required
            value={this.state.formData.holidays || ''}
          />
          <p className={styles.holidaysInfoText}>
            {t('settings:text.holidayInfoText')}
          </p>
        </form>
      </Modal>
    );
  }

  renderEditYearModal() {
    const {
      isPosting,
      t,
    } = this.props;

    return (
      <Modal
        actions={[
          {
            clickHandler: this.editHandler,
            label: t('general:action.save'),
            loading: isPosting,
          },
        ]}
        closeHandler={this.closeEditYearDialog}
        title={t('settings:title.editYear')}
        translations={{ close: t('general:action.close') }}
      >
        <form>
          <TextField
            changeHandler={this.changeHandler}
            disabled
            error={this.state.formValidity.elements.year}
            fieldId="year"
            label={t('config:element.year')}
            required
            type="number"
            value={this.state.formData.year || ''}
          />
          <TextArea
            changeHandler={this.changeHandler}
            error={this.state.formValidity.elements.holidays}
            fieldId="holidays"
            label={t('config:element.holidays')}
            required
            value={this.state.formData.holidays || ''}
          />
          <p className={styles.holidaysInfoText}>
            {t('settings:text.holidayInfoText')}
          </p>
        </form>
      </Modal>
    );
  }

  render() {
    const {
      config,
      isFetching,
      t,
    } = this.props;
    const {
      showAddYearDialog,
      showEditYearDialog,
    } = this.state;

    return (
      <Layout
        loading={isFetching}
        title={t('settings:title.settings')}
      >
        <div className={styles.yearsAndHolidaysHeaderBox}>
          <h2 className={styles.yearsAndHolidaysTitle}>
            {t('settings:title.yearsAndHolidays')}
          </h2>
          <div className={styles.yearButtonsWrapper}>
            <Button
              clickHandler={this.openAddYearDialog}
              icon="add"
              label={t('settings:action.addYear')}
              priority="primary"
            />
          </div>
        </div>
        {config && config.get('supportedYears').map(this.renderYearsAndHolidaysBox)}
        {showAddYearDialog && this.renderAddYearModal()}
        {showEditYearDialog && this.renderEditYearModal()}
      </Layout>
    );
  }
}

SettingsComponent.defaultProps = {
  config: null,
};

SettingsComponent.propTypes = {
  config: ImmutablePropTypes.mapContains({}),
  fetchConfig: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  isPosting: PropTypes.bool.isRequired,
  saveConfig: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
};

export default withNamespaces()(SettingsComponent);
