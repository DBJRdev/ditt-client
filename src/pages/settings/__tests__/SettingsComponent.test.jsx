import React from 'react';
import { shallow } from 'enzyme';
import SettingsComponent from '../SettingsComponent';
import configMock from '../../../../tests/mocks/configMock';

describe('rendering', () => {
  it('renders correctly', () => {
    const props = {
      config: configMock,
      fetchConfig: () => new Promise(() => {}),
      isFetching: false,
      isPosting: false,
      saveConfig: () => new Promise(() => {}),
    };
    const tree = shallow(<SettingsComponent {...props} />);

    expect(tree).toMatchSnapshot();
  });
});
