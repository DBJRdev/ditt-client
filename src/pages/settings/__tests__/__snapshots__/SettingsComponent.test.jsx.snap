// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<Connect(LayoutComponent)
  loading={false}
  title="settings:title.settings"
>
  <div
    className="yearsAndHolidaysHeaderBox"
  >
    <h2
      className="yearsAndHolidaysTitle"
    >
      settings:title.yearsAndHolidays
    </h2>
    <div
      className="yearButtonsWrapper"
    >
      <ForwardRef(withForwardedRef(WithProviderContextComponent))
        beforeLabel={
          <Icon
            icon="add"
            size="medium"
          />
        }
        label="settings:action.addYear"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="yearsAndHolidaysHeader"
  >
    <div
      className="yearsAndHolidaysHeaderBox"
    >
      <h3
        className="yearTitle"
      >
        2018
      </h3>
      <div
        className="yearButtonsWrapper"
      >
        <div
          className="yearEditButtonWrapper"
        >
          <ForwardRef(withForwardedRef(WithProviderContextComponent))
            beforeLabel={
              <Icon
                icon="edit"
                size="medium"
              />
            }
            label="settings:action.editYear"
            labelVisibility="none"
            onClick={[Function]}
            priority="outline"
          />
        </div>
        <ForwardRef(withForwardedRef(WithProviderContextComponent))
          beforeLabel={
            <Icon
              icon="expand_more"
              size="medium"
            />
          }
          label=""
          labelVisibility="none"
          onClick={[Function]}
          priority="outline"
        />
      </div>
    </div>
  </div>
  <div
    className="yearsAndHolidaysHeader"
  >
    <div
      className="yearsAndHolidaysHeaderBox"
    >
      <h3
        className="yearTitle"
      >
        2019
      </h3>
      <div
        className="yearButtonsWrapper"
      >
        <div
          className="yearEditButtonWrapper"
        >
          <ForwardRef(withForwardedRef(WithProviderContextComponent))
            beforeLabel={
              <Icon
                icon="edit"
                size="medium"
              />
            }
            label="settings:action.editYear"
            labelVisibility="none"
            onClick={[Function]}
            priority="outline"
          />
        </div>
        <ForwardRef(withForwardedRef(WithProviderContextComponent))
          beforeLabel={
            <Icon
              icon="expand_more"
              size="medium"
            />
          }
          label=""
          labelVisibility="none"
          onClick={[Function]}
          priority="outline"
        />
      </div>
    </div>
  </div>
  <div
    className="yearsAndHolidaysHeader"
  >
    <div
      className="yearsAndHolidaysHeaderBox"
    >
      <h3
        className="yearTitle"
      >
        2020
      </h3>
      <div
        className="yearButtonsWrapper"
      >
        <div
          className="yearEditButtonWrapper"
        >
          <ForwardRef(withForwardedRef(WithProviderContextComponent))
            beforeLabel={
              <Icon
                icon="edit"
                size="medium"
              />
            }
            label="settings:action.editYear"
            labelVisibility="none"
            onClick={[Function]}
            priority="outline"
          />
        </div>
        <ForwardRef(withForwardedRef(WithProviderContextComponent))
          beforeLabel={
            <Icon
              icon="expand_more"
              size="medium"
            />
          }
          label=""
          labelVisibility="none"
          onClick={[Function]}
          priority="outline"
        />
      </div>
    </div>
  </div>
  <div
    className="yearsAndHolidaysHeader"
  >
    <div
      className="yearsAndHolidaysHeaderBox"
    >
      <h3
        className="yearTitle"
      >
        2021
      </h3>
      <div
        className="yearButtonsWrapper"
      >
        <div
          className="yearEditButtonWrapper"
        >
          <ForwardRef(withForwardedRef(WithProviderContextComponent))
            beforeLabel={
              <Icon
                icon="edit"
                size="medium"
              />
            }
            label="settings:action.editYear"
            labelVisibility="none"
            onClick={[Function]}
            priority="outline"
          />
        </div>
        <ForwardRef(withForwardedRef(WithProviderContextComponent))
          beforeLabel={
            <Icon
              icon="expand_more"
              size="medium"
            />
          }
          label=""
          labelVisibility="none"
          onClick={[Function]}
          priority="outline"
        />
      </div>
    </div>
  </div>
</Connect(LayoutComponent)>
`;
